pkg/executor/adapter.go

```diff
@@ -186,16 +199,27 @@ func (a *recordSet) NewChunk(alloc chunk.Allocator) *chunk.Chunk {
 
 func (a *recordSet) Finish() error {
 	var err error
-	a.once.Do(func() {
-		err = exec.Close(a.executor)
-		cteErr := resetCTEStorageMap(a.stmt.Ctx)
-		if cteErr != nil {
-			logutil.BgLogger().Error("got error when reset cte storage, should check if the spill disk file deleted or not", zap.Error(cteErr))
-		}
-		if err == nil {
-			err = cteErr
-		}
-	})
+	if a.finishLock.TryLock() {
+		defer a.finishLock.Unlock()
+		a.once.Do(func() {
+			err = exec.Close(a.executor)
+			cteErr := resetCTEStorageMap(a.stmt.Ctx)
+			if cteErr != nil {
+				logutil.BgLogger().Error("got error when reset cte storage, should check if the spill disk file deleted or not", zap.Error(cteErr))
+			}
+			if err == nil {
+				err = cteErr
+			}
+			a.executor = nil
+			if a.stmt != nil {
+				status := a.stmt.Ctx.GetSessionVars().SQLKiller.GetKillSignal()
+				inWriteResultSet := a.stmt.Ctx.GetSessionVars().SQLKiller.InWriteResultSet.Load()
+				if status > 0 && inWriteResultSet {
+					logutil.BgLogger().Warn("kill query, this SQL may be stuck in the network I/O stack.", zap.Uint64("conn", a.stmt.Ctx.GetSessionVars().ConnectionID))
```

Comment:
will `kill connection` trigger this log?
if so, how do we distinguish `kill query` and `kill connection` from the log?